posts/2025-10-22-routing-in-angular-20-basics-to-navigation-guards.adoc

```diff
@@ -0,0 +1,348 @@
+
+When you need to route from code, favor clarity and explicitness.
+
+[source,typescript]
+----
+// features/auth/login.component.ts (snippet)
+import { Component, inject } from '@angular/core';
+import { ActivatedRoute, Router } from '@angular/router';
+import { AuthService } from './auth.service';
+
+@Component({
+  standalone: true,
+  template: `
+    <form (ngSubmit)="login()">
+      <!-- form fields -->
+      <button type="submit">Sign in</button>
+    </form>
+  `,
+})
+export class LoginComponent {
+  private router = inject(Router);
+  private route = inject(ActivatedRoute);
+  private auth = inject(AuthService);
+
+  async login() {
+    await this.auth.signIn();
+    const redirect = this.route.snapshot.queryParamMap.get('redirect') || '/dashboard';
+    this.router.navigateByUrl(redirect, { replaceUrl: true, state: { from: 'login' } });
+  }
+}
+----
+
+- navigate and navigateByUrl are both fine; navigate works with command arrays and relative routes.
+- replaceUrl avoids filling history with transient steps like login.
+
+=== Route guards: protect, match, and confirm
+
+Functional guards are terse, testable, and DI-friendly. Use canActivate for already-loaded routes and canMatch to gate access before lazy-loading. Use canDeactivate to protect against losing changes.
+
+[source,typescript]
+----
+// app/auth/guards.ts
+import { inject } from '@angular/core';
+import { CanActivateFn, CanDeactivateFn, CanMatchFn, Router, UrlSegment, Route } from '@angular/router';
+import { AuthService } from './auth.service';
+
+export const authCanActivate: CanActivateFn = (_route, state) => {
+  const auth = inject(AuthService);
+  const router = inject(Router);
+  if (auth.isLoggedIn()) return true;
+  // Return a UrlTree to redirect without throwing
+  return router.createUrlTree(['/login'], { queryParams: { redirect: state.url } });
+};
+
+export const authCanMatch: CanMatchFn = (route: Route, segments: UrlSegment[]) => {
+  const auth = inject(AuthService);
+  const router = inject(Router);
+  if (auth.isLoggedIn()) return true;
+
+  const attempted = '/' + segments.map(s => s.path).join('/');
+  return router.createUrlTree(['/login'], { queryParams: { redirect: attempted } });
+};
+
+export const unsavedCanDeactivate: CanDeactivateFn<{ hasUnsavedChanges(): boolean }> = (component) => {
+  return component.hasUnsavedChanges()
+    ? confirm('You have unsaved changes. Leave this page?')
+    : true;
+};
+----
+
+Notes:
+- Returning a UrlTree is the most ergonomic way to redirect from guards.
+- Prefer canMatch for lazy-loaded feature entries. It prevents loading code the user can’t access.
+
+=== Async redirects in Angular 20
+
+Redirects are often conditional and data-driven. Angular 20 refines ergonomics for async redirects by allowing redirect functions to return a Promise or Observable of a UrlTree, keeping logic declarative in the route table. When you need DI, the function executes in a proper injection context.
```

Comment:
I would use Dependency injection instead of the acronym DI as it was not explained before. You can use specified DI (Dependency Injection) above where you use it the first time and keep DI here if you prefer.